File: CF_Quick-One/920_div3/F.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;
#define int long long

int32_t main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;

    while (t--) {
        int N, Q;
        cin >> N >> Q;

        vector<int> Arr(N);
        for (int i = 0; i < N; ++i) {
            cin >> Arr[i];
        }

        int K = sqrt(N) + 1;

        // MAKE DP OF SIZE N X K
        long long dp[N][K];

        for (int k = 1; k < K; ++k) {
            for (int i = N - 1; i >= 0; --i) {
                if (i + k < N) {
                    dp[i][k] = dp[i + k][k] + Arr[i];
                } else {
                    dp[i][k] = Arr[i];
                }
            }
        }

        // making it a suffix
        for (int k = 1; k < K; ++k) {
            for (int i = N - 1; i >= 0; --i) {
                if (i + k < N) {
                    dp[i][k] += dp[i + k][k];
                }
            }
        }

        vector<long long> ANS;
        for (int q = 0; q < Q; ++q) {
            int s, delta, k;
            cin >> s >> delta >> k;
            s -= 1;

            if (delta < K) {
                long long temp = 0;
                temp += dp[s][delta];
                if (s + delta * k < N) {
                    temp -= dp[s + delta * k][delta];
                    long long x = dp[s + delta * k][delta];
                    if (s + delta * (k + 1) < N) {
                        x -= dp[s + delta * (k + 1)][delta];
                    }
                    temp -= k * x;
                }
                ANS.push_back(temp);
            } else {
                long long temp = 0;
                for (int i = 1; i <= k; ++i) {
                    temp += Arr[s + (i - 1) * delta] * i;
                }
                ANS.push_back(temp);
            }
        }

        for (long long ans : ANS) {
            cout << ans << " ";
        }
        cout << "\n";
    }

    return 0;
}

File: CF_Quick-One/920_div3/F.py
from sys import stdin
def input(): return stdin.readline()[:-1]


def solve():
    N, Q = map(int, input().split())
    Arr = list(map(int, input().split()))
    
    K = int(N ** 0.5) + 1
    
    # MAKE DP OF SIZE N X K
    dp = [[0 for _ in range(N)] for _ in range(K)]
    
    for k in range(1, K):
        for i in range(N-1, -1, -1):
            if i + k < N:
                dp[k][i] = dp[k][i+k] + Arr[i]
            else:
                dp[k][i] = Arr[i]
                
    # making it a suffix
    for k in range(1, K):
        for i in range(N-1, -1, -1):
            if i + k < N:
                dp[k][i] += dp[k][i+k]
    
    ANS = [] 
    for _ in range(Q):
        s, delta, k = map(int, input().split())
        s -= 1
        
        if delta < K:
            temp = 0
            temp += dp[delta][s]
            if s + delta*(k) < N:
                temp -= dp[delta][s+delta*(k)]
                x = dp[delta][s+delta*(k)]
                if (s + delta*(k+1)) < N:
                    x -= dp[delta][s+delta*(k+1)]
                temp -= k * x
            ANS.append(temp)
        else:
            temp = 0
            for i in range(1, k+1):
                temp += Arr[s + (i-1) * delta] * i
            ANS.append(temp)
            
    
    print(*ANS)
for _ in range(int(input())):
    solve()